d3a_interface/sim_results/kpi.py

```diff
@@ -162,6 +201,15 @@ def __repr__(self):
     def area_performance_indices(self, area_dict, core_stats):
         if area_dict['uuid'] not in self.state:
             self.state[area_dict['uuid']] = KPIState()
+
+        # initialization of house saving state
+        if area_dict['uuid'] not in self.saving_state and \
+                has_no_grand_child(area_dict):
+            self.saving_state[area_dict['uuid']] = SavingKPI()
+        if area_dict['uuid'] in self.saving_state:
```

Comment:
```suggestion
        if area_dict['uuid'] in self.saving_state:
          ...
        elif  has_no_grand_child(area_dict):
          ...
```

Reply:
Can you please explain it, sorry haven'T actually understood it

Reply:
This is to save up the first condition `area_dict['uuid'] not in self.saving_state`

Reply:
Target of saving up first condition will make loose saving kpi to be calculated for the first slot.

Reply:
> In our case I see that the odds of area_dict['uuid'] in self.saving_state returning True is high + this is a O(1) statement that will be executed very quickly unlike the has_no_grand_children(area_dict) which is a recursive call and this is why this should be placed second

So to summarize, you proposal is dependent on probability, meaning you are willing to skip calculation for the first time result service has started/rebooted?

How about creating a nested if clause to avoid calculation of the expensive  `has_no_grand_children(area_dict)`

```
    if area_dict['uuid'] not in self.saving_state:
        if has_no_grand_children(area_dict):
            self.saving_state[area_dict['uuid']] = SavingsKPI()
```